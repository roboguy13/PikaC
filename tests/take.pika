%generate take

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

NatLayout : layout[-x](Nat);
NatLayout Z := emp;
NatLayout (S n) :=
  exists +nxt.
  x :-> nxt **
  NatLayout n [nxt];

take : NatLayout -> Sll -> Sll;
take Z Nil := Nil [Sll];
take (S n) Nil := Nil [Sll];
take Z (Cons h t) := Nil [Sll];
take (S n) (Cons h t) :=
  Cons h (take n t) [Sll];

%test "take" Sll:
  take (S (S (Z [NatLayout]) [NatLayout]) [NatLayout])
    (Cons 10 (Cons 20 (Cons 30 (Cons 40 (Nil [Sll]) [Sll]) [Sll]) [Sll]) [Sll]);

