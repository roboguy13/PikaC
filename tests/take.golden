void take(loc x1, loc x, loc x2);
void take(loc x1, loc x, loc x2) {
 if (x == 0) {
  if (x1 == 0) {
   WRITE_LOC(x2, 0, 0);
  } else {
   WRITE_LOC(x2, 0, 0);
  }
 } else {
  loc h = READ_LOC(x, 0);
  loc nx = READ_LOC(x, 1);
  if (x1 == 0) {
   WRITE_LOC(x2, 0, 0);
  } else {
   loc n = READ_LOC(x1, 0);
   take(n, nx, x2);
   loc x4 = READ_LOC(x2, 0);
   loc x = (loc)malloc(2 * sizeof(loc));
   WRITE_LOC(x2, 0, x);
   WRITE_LOC(x, 1, x4);
   WRITE_LOC(x, 0, h);
  }
 }
}